sessionctx/context.go

```diff
@@ -149,3 +151,20 @@ const (
 	// LastExecuteDDL is the key for whether the session execute a ddl command last time.
 	LastExecuteDDL basicCtxType = 3
 )
+
+// ValidateReadTS validates that readTS does not exceed the current timestamp
+func ValidateReadTS(ctx context.Context, sctx Context, readTS uint64) error {
+	txnScope := sctx.GetSessionVars().CheckAndGetTxnScope()
+	latestTS, err := sctx.GetStore().GetOracle().GetLowResolutionTimestamp(ctx, &oracle.Option{TxnScope: txnScope})
+	// If we fail to get latestTS or the readTS exceeds it, get a timestamp from PD to double check
+	if err != nil || readTS > latestTS {
+		currentVer, err := sctx.GetStore().CurrentVersion(txnScope)
```

Comment:
How about use `GetOracle().GetStaleTimestamp(ctx, txnScope, 0)`

Reply:
Can I use `GetOracle().GetStaleTimestamp(ctx, txnScope, 0)` to replace L158, which is a fast path that does not involve an RPC?

Reply:
I'm afraid there is some risk. The timestamp from `GetStaleTimestamp` is calculated from the local time. If there is some clock drift, the timestamp may exceed the max ts in PD and still pass the validation. Then, it will break transactions.

Reply:
Does tikv/tikv#10443 already make the stale read not break a transaction in TiKV? Validate in TiDB should be pre-check and make the request fail fast. If you think that isn't a good way, maybe add metrics about how many request will validate from PD.

Reply:
I hope to reuse it to also check for `tidb_snapshot`. It is not a stale read request, so it may still advance the `max_ts` in TiKV and break a transaction.

Anyway, it is a good idea to add metrics for it.